crates/common/tedge_config/src/tedge_config_cli/new.rs

```diff
@@ -0,0 +1,655 @@
+                    let (tables, field) = target.rsplit_once('.').unwrap();
+                    for key in tables.split('.') {
+                        let table = doc.as_table_mut().unwrap();
+                        if !table.contains_key(key) {
+                            table.insert(key.to_owned(), toml::Value::Table(Table::new()));
+                        }
+                        doc = &mut doc[key];
+                    }
+                    let table = doc.as_table_mut().unwrap();
+                    // TODO if this returns Some, something is going wrong? Maybe this could be an error, or maybe it doesn't matter
+                    table.insert(field.to_owned(), value);
+                }
+            }
+            TomlMigrationStep::UpdateFieldValue { key, value } => {
+                let mut doc = &mut toml;
+                let (tables, field) = key.rsplit_once('.').unwrap();
+                for key in tables.split('.') {
+                    let table = doc.as_table_mut().unwrap();
+                    if !table.contains_key(key) {
+                        table.insert(key.to_owned(), toml::Value::Table(Table::new()));
+                    }
+                    doc = &mut doc[key];
+                }
+                let table = doc.as_table_mut().unwrap();
+                // TODO if this returns Some, something is going wrong? Maybe this could be an error, or maybe it doesn't matter
+                table.insert(field.to_owned(), value);
+            }
+            TomlMigrationStep::RemoveTableIfEmpty { key } => {
+                let mut doc = &mut toml;
+                let (parents, target) = key.rsplit_once('.').unwrap();
+                for key in parents.split('.') {
+                    let table = doc.as_table_mut().unwrap();
+                    if !table.contains_key(key) {
+                        table.insert(key.to_owned(), toml::Value::Table(Table::new()));
+                    }
+                    doc = &mut doc[key];
+                }
+                let table = doc.as_table_mut().unwrap();
+                if let Some(table) = table.get(target) {
+                    let table = table.as_table().unwrap();
+                    // TODO make sure this is covered in toml migration test
+                    if !table.is_empty() {
+                        return toml;
+                    }
+                }
+                table.remove(target);
+            }
+        }
+
+        toml
+    }
+}
+
+/// The keys that can be read from the configuration
+pub static READABLE_KEYS: Lazy<Vec<(Cow<'static, str>, doku::Type)>> = Lazy::new(|| {
+    let ty = TEdgeConfigReader::ty();
+    let doku::TypeKind::Struct { fields, transparent: false } = ty.kind else { panic!("Expected struct but got {:?}", ty.kind) };
+    let doku::Fields::Named { fields } = fields else { panic!("Expected named fields but got {:?}", fields)};
+    struct_field_paths(None, &fields)
+});
+
+impl TEdgeTomlVersion {
```

Comment:
So you propose to manage config versions and to have an automatic conversion. It's a good idea. We will have to double check with @reubenmiller for the proposed changes, though.

My questions are more about how this works.
- In a migration step like `mv("mqtt.port", MqttBindPort)`, what is this `MqttBindPort`. I failed to find a definition in the code.
- Do we have to maintain two versions of the config. In the current state of the PR, we have both but what are the plan?
-  and do we have to maintain

Reply:
`MqttBindPort` is `WritableKey::MqttBindPort`. `WritableKey` is an enum generated by `define_tedge_config!`, and I imported `WritableKey::*` in this method to reduce noise.

The idea is that the configuration will be read, the version will be checked, and then migrations carried out if necessary and written back to tedge.toml. If we've applied any migrations, we re-read the configuration before creating the `TEdgeConfigReader` from it.

I'm not expecting there to be a great burden from maintaining support for existing configuration keys and outdated toml files in future tedge versions, but it should be pretty simple to remove the relevant code if we wish to. As far as I understand, this logic should continue to just work for people upgrading tedge from whatever version, and tedge config will continue to accept (but warn about) the deprecated keys (there are far fewer of those in this PR than the previous PR due to the original key structure being largely preserved here). It uses a `toml::Value` based representation to ensure that the migration doesn't corrupt any existing data if the version field somehow disappears from the config file.